### ISSUE TITLE
Example how to use mlrMBO package for mlr

### ISSUE THREAD

**Issue by jyk** [mlr-org/mlr#1735]
Hi!
Where can I find example of how to use mlrMBO package for bayesian optimization of tuning parameters of xgboost model fitted with mlr package?

**Comment by Stevo15025** [comment]
This is an example we have not built yet (see the issue in the mlr-tutorial [here](https://github.com/mlr-org/mlr-tutorial/issues/102))

Hopefully soon!

**Comment by Stevo15025** [comment]
I did not work on mlrMBO, but here is a simple example that I believe covers a very basic approach. For more info I would suggest reading over the [paper](https://arxiv.org/abs/1703.03373) and blog [post](http://mlr-org.github.io/First_release_of_mlrMBO_the_toolbox_for_Bayesian_Black_Box_Optimization/).

@berndbischl @jakob-r @jakobbossek @mllg  is this about right?
```r
# Load data, create tsk and learner
data(BostonHousing, package = "mlbench")
tsk = makeRegrTask("Xgboost with Boston Housing", data = BostonHousing, target = "medv")
tsk = createDummyFeatures(tsk)
# NOTE: You may want to decrease the threads here depending on your machine
lrn = makeLearner("regr.xgboost", nthread = 4)

# standard resampling and paramset
res = makeResampleDesc("CV", iters = 5)
par = makeParamSet(
  makeIntegerParam(id = "nrounds", lower = 1,upper = 80),
  makeIntegerParam(id = "max_depth", lower = 2, upper = 15),
  makeNumericParam(id = "eta", lower = .01, upper = .4)
)
# This is mostly ripped and modified from the mbo blog post
library(mlrMBO)
# In this simple example we construct the control object with the defaults:
mbo.ctrl = makeMBOControl()
# For this numeric optimization we are going to use the Expected Improvement as infill criterion:
mbo.ctrl = setMBOControlInfill(mbo.ctrl, crit = crit.ei)
# We will allow for exactly 25 evaluations of the objective function:
mbo.ctrl = setMBOControlTermination(mbo.ctrl, max.evals = 25L)

# Make a design matrix 
design.mat = generateRandomDesign(n = 500, par.set = par)
# add the mbo control and design matrix to the mlrMBO tuning function
ctrl = makeTuneControlMBO(mbo.control = mbo.ctrl, mbo.design = design.mat)

# Everything else is pretty standard
tune.pars = tuneParams(learner = lrn, task = tsk, resampling = res,
              measures = rmse, par.set = par, control = ctrl)
```

**Comment by PhilippPro** [comment]
I think the easiest way is to read this blog post: [http://mlr-org.github.io/How-to-win-a-drone-in-20-lines-of-R-code/](http://mlr-org.github.io/How-to-win-a-drone-in-20-lines-of-R-code/)

Not more than 20 lines. ;)

**Comment by jyk** [comment]
Many thanks!
Jüri

**Comment by jyk** [comment]
I executed model based optimization but got error:

# Data preparation ***************************************

if (!require("AppliedPredictiveModeling")) {
  install.packages("AppliedPredictiveModeling", dependencies = TRUE)
  library(AppliedPredictiveModeling)
}
set.seed(100)
if (!require("caret")) {
  install.packages("caret", dependencies = TRUE)
  library(caret)
}
data(GermanCredit)
# First, remove near-zero variance predictors then get rid of a few predictors
# that duplicate values. For example, there are two possible values for the
# housing variable: "Rent", "Own" and "ForFree". So that we don't have linear
# dependencies, we get rid of one of the levels (e.g. "ForFree")
GermanCredit <- GermanCredit[, -nearZeroVar(GermanCredit)]
GermanCredit$CheckingAccountStatus.lt.0 <- NULL
GermanCredit$SavingsAccountBonds.lt.100 <- NULL
GermanCredit$EmploymentDuration.lt.1 <- NULL
GermanCredit$EmploymentDuration.Unemployed <- NULL
GermanCredit$Personal.Male.Married.Widowed <- NULL
GermanCredit$Property.Unknown <- NULL
GermanCredit$Housing.ForFree <- NULL
# Rename case as "Yes" and non-case as "No"
GermanCredit$Class <- ifelse(GermanCredit$Class == "Bad", "Yes", "No")
levels(GermanCredit$Class) <- list(no="No", yes="Yes")
table(GermanCredit$Class)
GermanCredit$Class <- factor(GermanCredit$Class)
#head(GermanCredit)
inTrain <- createDataPartition(GermanCredit$Class, p = .8)[[1]]
DF.train <- GermanCredit[ inTrain, ]
DF.test  <- GermanCredit[-inTrain, ]

# Modeling ***************************************

library(mlr)
library(mlrMBO)
library(parallelMap)

# Setting learner
lrn <- makeLearner("classif.xgboost", eval_metric = "auc", predict.type = "prob")

# Setting hyperparameters' space
ps <- makeParamSet(
          makeIntegerParam("nrounds", lower = 200, upper = 2500, default = 200),
          makeNumericParam("eta", lower = -7, upper = -5, default = -6, 
                           trafo = function(x) 2^x),
          makeIntegerParam("max_depth", lower = 3, upper = 15, default = 3),
          makeNumericParam("colsample_bytree", lower = 0.3, upper = 1, default = 0.6),
          makeNumericParam("subsample", lower = 0.3, upper = 1, default = 0.6)
          )

# Setting classification task
task <- makeClassifTask(data = DF.train, target = "Class", positive = "Yes")
task

# Setting control object for MBO optimization  - Bayesian optimization (aka model based optimization)
mbo.ctrl <- makeMBOControl(save.on.disk.at = c(0, 5, 10, 20, 50, 75, 85, 95))

# Extends an MBO control object with infill criteria and infill optimizer options
mbo.ctrl <- setMBOControlTermination(mbo.ctrl, iters = 100)

# Defining surrogate learner
surrogate.lrn <- makeLearner("regr.km", predict.type = "se")

# Create control object for hyperparameter tuning with MBO
ctrl = mlr:::makeTuneControlMBO(learner = surrogate.lrn, mbo.control = mbo.ctrl)

# Tuning hyperparameters
parallelStartMulticore()
res.mbo = tuneParams(learner=lrn, 
                     task=task, 
                     resampling=cv10, 
                     par.set = ps, 
                     control = ctrl, 
                     show.info = TRUE, 
                     measures = auc
                     )
parallelStop()



> res.mbo = tuneParams(learner=lrn, 
+                      task=task, 
+                      resampling=cv10, 
+                      par.set = ps, 
+                      control = ctrl, 
+                      show.info = TRUE, 
+                      measures = auc
+                      )
[Tune] Started tuning learner classif.xgboost for parameter set:
                    Type len Def         Constr Req Tunable Trafo
nrounds          integer   - 200 200 to 2.5e+03   -    TRUE     -
eta              numeric   -  -6       -7 to -5   -    TRUE     Y
max_depth        integer   -   3        3 to 15   -    TRUE     -
colsample_bytree numeric   - 0.6       0.3 to 1   -    TRUE     -
subsample        numeric   - 0.6       0.3 to 1   -    TRUE     -
With control class: TuneControlMBO
Imputation value: -0
Warning in checkStuff(fun, par.set, design, learner, control) :
  You turned off the final saving of the optimization result at (iter + 1)! Do you really want this?
[Tune-x] 1: nrounds=1206; eta=0.0178; max_depth=12; colsample_bytree=0.771; subsample=0.461
Error in UseMethod("predict") : 
  no applicable method for 'predict' applied to an object of class "xgb.Booster"

**Comment by jyk** [comment]
In addition, for  Stevo15025's example I have got error message for line 
# Everything else is pretty standard
tune.pars = tuneParams(learner = lrn, task = tsk, resampling = res,
                       measures = rmse, par.set = par, control = ctrl
                       )

:

[Tune] Started tuning learner regr.xgboost for parameter set:
             Type len Def      Constr Req Tunable Trafo
nrounds   integer   -   -     1 to 80   -    TRUE     -
max_depth integer   -   -     2 to 15   -    TRUE     -
eta       numeric   -   - 0.01 to 0.4   -    TRUE     -
With control class: TuneControlMBO
Imputation value: Inf
[Tune-x] 1: nrounds=40; max_depth=9; eta=0.134
Error in UseMethod("predict") : 
  no applicable method for 'predict' applied to an object of class "xgb.Booster"

**Comment by Stevo15025** [comment]
Do you have the most up to date version of xgboost?

**Comment by jyk** [comment]
Yes, I had too old version. Now with 0.6.4 it is working. However, I have got next warning message: 
Warning in checkStuff(fun, par.set, design, learner, control) :
  You turned off the final saving of the optimization result at (iter + 1)! Do you really want this?

What should be done and what does it mean ?
Jüri

**Comment by PhilippPro** [comment]
No need to worry, I think your final mbo model is only saved at 0, 5, 10, 20, 50, 75, 85, 95 (like defined in mbo.ctrl <- makeMBOControl(save.on.disk.at = c(0, 5, 10, 20, 50, 75, 85, 95))), but you will get results. The documentation should be better here @jakob-r, I also asked myself this question some time ago.

**Comment by jakob-r** [comment]
I will try to improve the documentation without getting to wordy. But @PhilippPro is right, It is recommended to save your run on the disk also after the last iteration because that is where you usually want to continue your search, when you are not yet satisfied with the result. The other stages are usually just used on a cluster where your process might be killed because it runs into a wall time.

**Comment by jyk** [comment]
Ok, many thanks!
